### PR TITLE
feat: allow admins to modify notification template delivery method

### PR DESCRIPTION
**Pull request by dannykopping**
Part of https://github.com/coder/team-coconut/issues/19

Admins must be able to modify which delivery method is set for each notification template. For all templates with a method not set, it falls back to the deployment-wide setting.

This PR enables the above functionality, and modifies the notification dispatch process to take custom dispatch methods per template into account.

Added two new routes:

- `PUT /notifications/templates/{notification_template}/method`
  - Enterprise only, allows admins to modify the delivery method on each notification template
  - Enables use-cases like "send all notifications via email, but send the _user account created_ notification via webhook so I can programmatically update my billing registry"
- `GET /notifications/templates/system`
  - AGPL, retrieves all notification templates of kind `system`
  - this is the only kind right now; in the future we'll have user-defined notification templates

Also made some small drive-by changes to the _notification settings_ work recently implemented.

### CODE REVIEW

**Comment by dannykopping** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/coder/coder/14116?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#14117** <a href="https://app.graphite.dev/github/pr/coder/coder/14117?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#14116** <a href="https://app.graphite.dev/github/pr/coder/coder/14116?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#14100** <a href="https://app.graphite.dev/github/pr/coder/coder/14100?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#14097](https://github.com/coder/coder/pull/14097) <a href="https://app.graphite.dev/github/pr/coder/coder/14097?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @dannykopping and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: coderd/notifications/enqueuer.go [hunk 1]
```diff
@@ -37,17 +34,31 @@ func NewStoreEnqueuer(cfg codersdk.NotificationsConfig, store Store, helpers tem
...
+		NotificationTemplateID: templateID,
+	})
+	if err != nil {
+		s.log.Warn(ctx, "failed to fetch message metadata", slog.F("template_id", templateID), slog.F("user_id", userID), slog.Error(err))
+		return nil, xerrors.Errorf("new message metadata: %w", err)
+	}
+
+	dispatchMethod := s.defaultMethod
+	if metadata.CustomMethod.Valid {
+		dispatchMethod = metadata.CustomMethod.NotificationMethod
```
Comment (dannykopping) [inline coderd/notifications/enqueuer.go:58]: Here is where the new dispatch method is chosen

File: cli/notifications_test.go [hunk 2]
```diff
@@ -16,6 +16,16 @@ import (
 	"github.com/coder/coder/v2/testutil"
 )
 
+func createOpts(t *testing.T) *coderdtest.Options {
```
Comment (johnstcn) [inline cli/notifications_test.go:19]: suggestion:

```
func NewWithDatabase(t testing.TB, options *Options, muts ...func(*Options)) { ... }

func WithExperiments(es ...codersdk.Experiment) func(*Options) { ... }

// then:
ownerClient, db := coderdtest.NewWithDatabase(t, nil, coderdtest.WithExperiments(codersdk.ExperimentNotifications))
```

File: coderd/notifications.go [hunk 3]
```diff
@@ -110,13 +102,58 @@ func (api *API) putNotificationsSettings(rw http.ResponseWriter, r *http.Request
 	}
 
 	err = api.Database.UpsertNotificationsSettings(ctx, string(settingsJSON))
+	if err != nil {
+		if rbac.IsUnauthorizedError(err) {
+			httpapi.Forbidden(rw)
```
Comment (johnstcn) [inline coderd/notifications.go:107]: suggestion: early return after here and then you can de-dent the else

**Review by johnstcn — APPROVED** [review]

File: cli/notifications_test.go (hunk 2 above)
Comment (dannykopping) [reply cli/notifications_test.go:19]: Yeah that would be nicer indeed.
I'll defer that to a subsequent PR though if you're ok with that?

File: coderd/notifications.go (hunk 3 above)
Comment (dannykopping) [reply coderd/notifications.go:107]: Good call 👍

File: cli/notifications_test.go (hunk 2 above)
Comment (johnstcn) [reply cli/notifications_test.go:19]: 💯

**Pull request merged**